soda/core/soda/execution/data_source.py

```diff
@@ -488,26 +488,41 @@ def profiling_sql_aggregates_text(self, table_name: str, column_name: str) -> st
         )
 
     def histogram_sql_and_boundaries(
-        self, table_name: str, column_name: str, min: int | float, max: int | float
+        self,
+        table_name: str,
+        column_name: str,
+        minimum: int | float,
+        maximum: int | float,
+        n_distinct: int,
+        column_type: str,
     ) -> tuple[str | None, list[int | float]]:
         # TODO: make configurable or derive dynamically based on data quantiles etc.
-        number_of_bins: int = 20
+        max_n_bins = 20
+        number_of_bins: int = max(1, min(n_distinct, max_n_bins))
+        number_of_intervals: int = number_of_bins - 1
 
-        if not min < max:
+        if not minimum < maximum:
             self.logs.warning(
-                f"Min of {column_name} on table: {table_name} must be smaller than max value. Min is {min}, and max is {max}"
+                f"Min of {column_name} on table: {table_name} must be smaller than max value. Min is {minimum}, and max is {maximum}"
             )
             return None, []
 
-        min_value = floor(min * 1000) / 1000
-        max_value = ceil(max * 1000) / 1000
-        bin_width = (max_value - min_value) / number_of_bins
+        min_value = floor(minimum * 1000) / 1000
+        max_value = ceil(maximum * 1000) / 1000
+        if column_type == "integer":
+            min_value = int(min_value)
+            max_value = int(max_value)
+            bin_width = (max_value - min_value) / number_of_intervals
+            if bin_width.is_integer():
+                bin_width = int(bin_width)
+        else:
+            bin_width = (max_value - min_value) / number_of_bins
```

Comment:
Hey Baturay, not sure whether this else statement should be here. Setting the `bin_width = (max_value - min_value) / number_of_bins` seems inconsistent with `for _ in range(0, number_of_intervals)` that is used later. 

If you have float data with `min_value` and `max_value`, `5` and `30`, respectively, then `number_of_bins` is `20`. In that case your `bin_width` would be set to `1.25`. In the for loop your bins would be set to `[5, 6.25, ..., 28.75]` whereas you would want `[5, 6.32, ..., 30]`.

Reply:
fixed